File: android/Reguerta/app/src/main/java/com/reguerta/user/presentation/shiftcoverage/ShiftCoverageViewModel.kt
package com.reguerta.user.presentation.shiftcoverage

import androidx.annotation.MainThread
import androidx.lifecycle.ViewModel
import com.reguerta.user.domain.shiftcoverage.ShiftCoverageCommand
import com.reguerta.user.domain.shiftcoverage.ShiftCoverageFailure
import com.reguerta.user.domain.shiftcoverage.ShiftCoverageRepository
import com.reguerta.user.domain.shiftcoverage.ShiftCoverageSession
import com.reguerta.user.domain.shiftcoverage.ShiftCoverageSnapshot
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow

internal data class ShiftCoverageState(
    val session: ShiftCoverageSession? = null,
    val snapshot: ShiftCoverageSnapshot? = null,
    val pendingCommand: ShiftCoverageCommand? = null,
    val failure: ShiftCoverageFailure? = null,
    val isBusy: Boolean = false,
)

/** Caller owns structured coroutines. All entry points and state changes belong to the UI dispatcher. */
@MainThread
internal class ShiftCoverageViewModel(private val repository: ShiftCoverageRepository) : ViewModel() {
    private val mutableState = MutableStateFlow(ShiftCoverageState())
    val state = mutableState.asStateFlow()
    private var generation = 0L
    private var receivedAtNanos = System.nanoTime()

    val nowMillis: Long
        get() = (state.value.snapshot?.serverTimeMillis ?: 0) + (System.nanoTime() - receivedAtNanos) / 1_000_000

    fun bind(session: ShiftCoverageSession?) {
        if (state.value.session == session) return
        generation++
        mutableState.value = ShiftCoverageState(session = session)
    }

    suspend fun refresh() {
        val session = state.value.session ?: return
        if (state.value.isBusy) return
        val owner = generation
        mutableState.value = state.value.copy(isBusy = true, snapshot = null, failure = null)
        try {
            val result = repository.read(null, session)
            if (owner != generation) return
            currentCoroutineContext().ensureActive()
            receivedAtNanos = System.nanoTime()
            mutableState.value = state.value.copy(snapshot = result)
        } catch (error: Exception) {
            if (owner != generation) return
            record(error)
        } finally {
            if (owner == generation) mutableState.value = state.value.copy(isBusy = false)
        }
    }

    suspend fun submit(command: ShiftCoverageCommand) {
        val current = state.value
        val session = current.session ?: return
        if (current.isBusy || current.snapshot == null || current.pendingCommand != null) return
        mutableState.value = current.copy(pendingCommand = command)
        executePending(session)
    }

    suspend fun retryPending() {
        val current = state.value
        val session = current.session ?: return
        if (current.isBusy || current.pendingCommand == null) return
        executePending(session)
    }

    private suspend fun executePending(session: ShiftCoverageSession) {
        val command = state.value.pendingCommand ?: return
        val owner = generation
        mutableState.value = state.value.copy(isBusy = true, snapshot = null, failure = null)
        try {
            repository.execute(command, session)
            if (owner != generation) return
            // A read-back failure cannot turn an acknowledged command back into an uncertain write.
            mutableState.value = state.value.copy(pendingCommand = null)
            currentCoroutineContext().ensureActive()
            val result = repository.read(null, session)
            if (owner != generation) return
            currentCoroutineContext().ensureActive()
            receivedAtNanos = System.nanoTime()
            mutableState.value = state.value.copy(snapshot = result)
        } catch (error: Exception) {
            if (owner != generation) return
            if (error is ShiftCoverageFailure.Rejected && error.status in listOf(400, 401, 403, 409)) {
                mutableState.value = state.value.copy(pendingCommand = null)
            }
            record(error)
        } finally {
            if (owner == generation) mutableState.value = state.value.copy(isBusy = false)
        }
    }

    private fun record(error: Exception) {
        val failure = error as? ShiftCoverageFailure ?: ShiftCoverageFailure.Unavailable
        if (failure == ShiftCoverageFailure.SessionChanged ||
            failure is ShiftCoverageFailure.Rejected && failure.status in listOf(401, 403)
        ) {
            bind(null)
        }
        mutableState.value = state.value.copy(failure = failure)
    }
}

File: ios/Reguerta/Reguerta/Presentation/ShiftCoverage/ShiftCoverageViewModel.swift
import Foundation
import Observation

/// Owns a single coverage inbox and serial mutation for one authorization revision.
/// A lost command response retains its exact intent for explicit replay; it never implies rollback.
@MainActor
@Observable
final class ShiftCoverageViewModel {
    private(set) var session: ShiftCoverageSession?
    private(set) var snapshot: ShiftCoverageSnapshot?
    private(set) var pendingCommand: ShiftCoverageCommand?
    private(set) var failure: ShiftCoverageFailure?
    private(set) var isBusy = false
    @ObservationIgnored private let repository: any ShiftCoverageRepository
    @ObservationIgnored private var receivedAt = ContinuousClock.now
    @ObservationIgnored private var detailEventId: String?
    @ObservationIgnored private var overviewRequested = false
    @ObservationIgnored private var generation: UInt64 = 0

    init(repository: any ShiftCoverageRepository) {
        self.repository = repository
    }

    func bind(_ session: ShiftCoverageSession?) {
        guard self.session != session else { return }
        generation &+= 1
        self.session = session
        detailEventId = nil
        overviewRequested = false
        snapshot = nil
        pendingCommand = nil
        failure = nil
        isBusy = false
    }

    var nowMillis: Int64 {
        guard let snapshot else { return 0 }
        let elapsed = receivedAt.duration(to: .now).components
        return snapshot.serverTimeMillis + elapsed.seconds * 1000 + elapsed.attoseconds / 1_000_000_000_000_000
    }

    func refresh() async {
        _ = await load(notificationEventId: detailEventId)
    }

    func refreshOverview() async {
        overviewRequested = true
        guard !isBusy else { return }
        overviewRequested = false
        detailEventId = nil
        _ = await load(notificationEventId: nil)
    }

    func openNotification(_ eventId: String) async -> String? {
        guard pendingCommand == nil else { return nil }
        return await load(notificationEventId: eventId)
    }

    private func load(notificationEventId: String?) async -> String? {
        guard let session, !isBusy else { return nil }
        let owner = generation
        isBusy = true
        snapshot = nil
        failure = nil
        var selectedCaseId: String?
        do {
            let result: ShiftCoverageSnapshot
            if let notificationEventId {
                result = try await repository.readNotification(eventId: notificationEventId, session: session)
            } else {
                result = try await repository.read(caseId: nil, session: session)
            }
            guard generation == owner else { return nil }
            try Task.checkCancellation()
            receivedAt = .now
            snapshot = result
            detailEventId = notificationEventId
            selectedCaseId = result.notification?.caseId
        } catch {
            guard generation == owner else { return nil }
            record(error)
        }
        guard generation == owner else { return nil }
        isBusy = false
        if overviewRequested {
            await refreshOverview()
            return nil
        }
        return selectedCaseId
    }

    func submit(_ command: ShiftCoverageCommand) async {
        guard let session, !isBusy, snapshot != nil, pendingCommand == nil else { return }
        pendingCommand = command
        await executePending(session: session)
    }

    func retryPending() async {
        guard let session, !isBusy, pendingCommand != nil else { return }
        await executePending(session: session)
    }

    private func executePending(session: ShiftCoverageSession) async {
        guard let command = pendingCommand else { return }
        let owner = generation
        isBusy = true
        snapshot = nil
        failure = nil
        do {
            try await repository.execute(command, session: session)
            guard generation == owner else { return }
            // An acknowledged write stays acknowledged even if read-back fails or the task is cancelled.
            pendingCommand = nil
            try Task.checkCancellation()
            let result: ShiftCoverageSnapshot
            if let detailEventId {
                result = try await repository.readNotification(eventId: detailEventId, session: session)
            } else {
                result = try await repository.read(caseId: nil, session: session)
            }
            guard generation == owner else { return }
            try Task.checkCancellation()
            receivedAt = .now
            snapshot = result
        } catch {
            guard generation == owner else { return }
            if case .rejected(let status, _) = error as? ShiftCoverageFailure,
               [400, 401, 403, 409].contains(status) {
                pendingCommand = nil
            }
            record(error)
        }
        guard generation == owner else { return }
        isBusy = false
        if overviewRequested {
            await refreshOverview()
        }
    }

    private func record(_ error: any Error) {
        failure = (error as? ShiftCoverageFailure) ?? .unavailable
        if failure == .sessionChanged {
            bind(nil)
            failure = .sessionChanged
        } else if case .rejected(let status, _) = failure, [401, 403].contains(status) {
            let rejection = failure
            bind(nil)
            failure = rejection
        }
    }
}
